// fp-ts/lib/NonEmptyArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Array_1 = require("./Array");
var function_1 = require("./function");
var Option_1 = require("./Option");
var Semigroup_1 = require("./Semigroup");
exports.URI = 'NonEmptyArray';
/**
 * Data structure which represents non-empty arrays
 * @data
 * @constructor NonEmptyArray
 * @since 1.0.0
 */
var NonEmptyArray = /** @class */ (function () {
    function NonEmptyArray(head, tail) {
        this.head = head;
        this.tail = tail;
    }
    /**
     * Converts this {@link NonEmptyArray} to plain {@link Array}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * assert.deepEqual(new NonEmptyArray(1, [2, 3]).toArray(), [1, 2, 3])
     */
    NonEmptyArray.prototype.toArray = function () {
        return function_1.concat([this.head], this.tail);
    };
    /**
     * Concatenates this {@link NonEmptyArray} and passed {@link Array}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * assert.deepEqual(new NonEmptyArray<number>(1, []).concatArray([2]), new NonEmptyArray(1, [2]))
     */
    NonEmptyArray.prototype.concatArray = function (as) {
        return new NonEmptyArray(this.head, function_1.concat(this.tail, as));
    };
    /**
     * Instance-bound implementation of {@link Functor}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * const double = (n: number): number => n * 2
     * assert.deepEqual(new NonEmptyArray(1, [2]).map(double), new NonEmptyArray(2, [4]))
     */
    NonEmptyArray.prototype.map = function (f) {
        return new NonEmptyArray(f(this.head), this.tail.map(f));
    };
    /**
     * Instance-bound implementation of {@link Apply}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * const x = new NonEmptyArray(1, [2])
     * const double = (n: number): number => n * 2
     * assert.deepEqual(x.ap(new NonEmptyArray(double, [double])).toArray(), [2, 4, 2, 4])
     */
    NonEmptyArray.prototype.ap = function (fab) {
        var _this = this;
        return fab.chain(function (f) { return _this.map(f); }); // <= derived
    };
    /**
     * Flipped version of {@link ap}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * const x = new NonEmptyArray(1, [2])
     * const double = (n: number) => n * 2
     * assert.deepEqual(new NonEmptyArray(double, [double]).ap_(x).toArray(), [2, 4, 2, 4])
     */
    NonEmptyArray.prototype.ap_ = function (fb) {
        return fb.ap(this);
    };
    /**
     * Instance-bound implementation of {@link Chain}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * const x = new NonEmptyArray(1, [2])
     * const f = (a: number) => new NonEmptyArray(a, [4])
     * assert.deepEqual(x.chain(f).toArray(), [1, 4, 2, 4])
     */
    NonEmptyArray.prototype.chain = function (f) {
        return f(this.head).concatArray(Array_1.array.chain(this.tail, function (a) { return f(a).toArray(); }));
    };
    /**
     * Instance-bound implementation of {@link Semigroup}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * const x = new NonEmptyArray(1, [2])
     * const y = new NonEmptyArray(3, [4])
     * assert.deepEqual(x.concat(y).toArray(), [1, 2, 3, 4])
     */
    NonEmptyArray.prototype.concat = function (y) {
        return this.concatArray(y.toArray());
    };
    /**
     * Instance-bound implementation of {@link Foldable}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * const x = new NonEmptyArray('a', ['b'])
     * assert.strictEqual(x.reduce('', (b, a) => b + a), 'ab')
     */
    NonEmptyArray.prototype.reduce = function (b, f) {
        return Array_1.array.reduce(this.toArray(), b, f);
    };
    /**
     * Instance-bound implementation of {@link Extend}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     * import { fold, monoidSum } from 'fp-ts/lib/Monoid'
     *
     * const sum = (as: NonEmptyArray<number>) => fold(monoidSum)(as.toArray())
     * assert.deepEqual(new NonEmptyArray(1, [2, 3, 4]).extend(sum), new NonEmptyArray(10, [9, 7, 4]))
     */
    NonEmptyArray.prototype.extend = function (f) {
        return unsafeFromArray(Array_1.array.extend(this.toArray(), function (as) { return f(unsafeFromArray(as)); }));
    };
    /**
     * Instance-bound implementation of {@link Comonad}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * assert.strictEqual(new NonEmptyArray(1, [2, 3]).extract(), 1)
     */
    NonEmptyArray.prototype.extract = function () {
        return this.head;
    };
    /**
     * Same as {@link toString}
     */
    NonEmptyArray.prototype.inspect = function () {
        return this.toString();
    };
    /**
     * Return stringified representation of this {@link NonEmptyArray}
     */
    NonEmptyArray.prototype.toString = function () {
        return "new NonEmptyArray(" + function_1.toString(this.head) + ", " + function_1.toString(this.tail) + ")";
    };
    /**
     * Gets minimum of this {@link NonEmptyArray} using specified {@link Ord} instance
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     * import { ordNumber } from 'fp-ts/lib/Ord'
     *
     * assert.strictEqual(new NonEmptyArray(1, [2, 3]).min(ordNumber), 1)
     *
     * @since 1.3.0
     */
    NonEmptyArray.prototype.min = function (ord) {
        return Semigroup_1.fold(Semigroup_1.getMeetSemigroup(ord))(this.head)(this.tail);
    };
    /**
     * Gets maximum of this {@link NonEmptyArray} using specified {@link Ord} instance
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     * import { ordNumber } from 'fp-ts/lib/Ord'
     *
     * assert.strictEqual(new NonEmptyArray(1, [2, 3]).max(ordNumber), 3)
     *
     * @since 1.3.0
     */
    NonEmptyArray.prototype.max = function (ord) {
        return Semigroup_1.fold(Semigroup_1.getJoinSemigroup(ord))(this.head)(this.tail);
    };
    /**
     * Gets last element of this {@link NonEmptyArray}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * assert.strictEqual(new NonEmptyArray(1, [2, 3]).last(), 3)
     * assert.strictEqual(new NonEmptyArray(1, []).last(), 1)
     *
     * @since 1.6.0
     */
    NonEmptyArray.prototype.last = function () {
        return Array_1.last(this.tail).getOrElse(this.head);
    };
    /**
     * Sorts this {@link NonEmptyArray} using specified {@link Ord} instance
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     * import { ordNumber } from 'fp-ts/lib/Ord'
     *
     * assert.deepEqual(new NonEmptyArray(3, [2, 1]).sort(ordNumber), new NonEmptyArray(1, [2, 3]))
     *
     * @since 1.6.0
     */
    NonEmptyArray.prototype.sort = function (ord) {
        return unsafeFromArray(Array_1.sort(ord)(this.toArray()));
    };
    /**
     * Reverts this {@link NonEmptyArray}
     *
     * @example
     * import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray'
     *
     * assert.deepEqual(new NonEmptyArray(1, [2, 3]).reverse(), new NonEmptyArray(3, [2, 1]))
     *
     * @since 1.6.0
     */
    NonEmptyArray.prototype.reverse = function () {
        return unsafeFromArray(this.toArray().reverse());
    };
    /**
     * @since 1.10.0
     */
    NonEmptyArray.prototype.length = function () {
        return 1 + this.tail.length;
    };
    return NonEmptyArray;
}());
exports.NonEmptyArray = NonEmptyArray;
var unsafeFromArray = function (as) {
    return new NonEmptyArray(as[0], as.slice(1));
};
/**
 * Builds {@link NonEmptyArray} from {@link Array} returning {@link Option#none} or {@link Option#some} depending on amount of values in passed array
 * @function
 * @since 1.0.0
 */
exports.fromArray = function (as) {
    return as.length > 0 ? Option_1.some(unsafeFromArray(as)) : Option_1.none;
};
var map = function (fa, f) {
    return fa.map(f);
};
var of = function (a) {
    return new NonEmptyArray(a, []);
};
var ap = function (fab, fa) {
    return fa.ap(fab);
};
var chain = function (fa, f) {
    return fa.chain(f);
};
var concat = function (fx, fy) {
    return fx.concat(fy);
};
/**
 * Builds {@link Semigroup} instance for {@link NonEmptyArray} of specified type arument
 * @function
 * @since 1.0.0
 */
exports.getSemigroup = function () {
    return { concat: concat };
};
/**
 * Group equal, consecutive elements of an array into non empty arrays.
 *
 * @example
 * import { NonEmptyArray, group } from 'fp-ts/lib/NonEmptyArray'
 * import { ordNumber } from 'fp-ts/lib/Ord'
 *
 * assert.deepEqual(group(ordNumber)([1, 2, 1, 1]), [
 *   new NonEmptyArray(1, []),
 *   new NonEmptyArray(2, []),
 *   new NonEmptyArray(1, [1])
 * ])
 *
 * @function
 * @since 1.7.0
 */
exports.group = function (S) { return function (as) {
    var r = [];
    var len = as.length;
    if (len === 0) {
        return r;
    }
    var head = as[0];
    var tail = [];
    for (var i = 1; i < len; i++) {
        var x = as[i];
        if (S.equals(x, head)) {
            tail.push(x);
        }
        else {
            r.push(new NonEmptyArray(head, tail));
            head = x;
            tail = [];
        }
    }
    r.push(new NonEmptyArray(head, tail));
    return r;
}; };
/**
 * Sort and then group the elements of an array into non empty arrays.
 *
 * @example
 * import { NonEmptyArray, groupSort } from 'fp-ts/lib/NonEmptyArray'
 * import { ordNumber } from 'fp-ts/lib/Ord'
 *
 * assert.deepEqual(groupSort(ordNumber)([1, 2, 1, 1]), [new NonEmptyArray(1, [1, 1]), new NonEmptyArray(2, [])])
 *
 * @function
 * @since 1.7.0
 */
exports.groupSort = function (O) {
    return function_1.compose(exports.group(O), Array_1.sort(O));
};
var reduce = function (fa, b, f) {
    return fa.reduce(b, f);
};
var foldMap = function (M) { return function (fa, f) {
    return fa.tail.reduce(function (acc, a) { return M.concat(acc, f(a)); }, f(fa.head));
}; };
var foldr = function (fa, b, f) {
    return f(fa.head, fa.tail.reduceRight(function (acc, a) { return f(a, acc); }, b));
};
var extend = function (fa, f) {
    return fa.extend(f);
};
var extract = function (fa) {
    return fa.extract();
};
function traverse(F) {
    var traverseF = Array_1.array.traverse(F);
    return function (ta, f) {
        var fb = f(ta.head);
        var fbs = traverseF(ta.tail, f);
        return F.ap(F.map(fb, function (b) { return function (bs) { return new NonEmptyArray(b, bs); }; }), fbs);
    };
}
function sequence(F) {
    var sequenceF = Array_1.array.sequence(F);
    return function (ta) {
        return F.ap(F.map(ta.head, function (a) { return function (as) { return new NonEmptyArray(a, as); }; }), sequenceF(ta.tail));
    };
}
/**
 * Splits an array into sub-non-empty-arrays stored in an object, based on the result of calling a `string`-returning
 * function on each element, and grouping the results according to values returned
 *
 * @example
 * import { NonEmptyArray, groupBy } from 'fp-ts/lib/NonEmptyArray'
 *
 * assert.deepEqual(groupBy(['foo', 'bar', 'foobar'], a => String(a.length)), {
 *   '3': new NonEmptyArray('foo', ['bar']),
 *   '6': new NonEmptyArray('foobar', [])
 * })
 *
 * @function
 * @since 1.10.0
 */
exports.groupBy = function (as, f) {
    var r = {};
    for (var _i = 0, as_1 = as; _i < as_1.length; _i++) {
        var a = as_1[_i];
        var k = f(a);
        if (r.hasOwnProperty(k)) {
            r[k].tail.push(a);
        }
        else {
            r[k] = new NonEmptyArray(a, []);
        }
    }
    return r;
};
/**
 * @instance
 * @since 1.0.0
 */
exports.nonEmptyArray = {
    URI: exports.URI,
    extend: extend,
    extract: extract,
    map: map,
    of: of,
    ap: ap,
    chain: chain,
    reduce: reduce,
    foldMap: foldMap,
    foldr: foldr,
    traverse: traverse,
    sequence: sequence
};
